ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,254 @@
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE EmptyCase             #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE StandaloneDeriving    #-}
+
+module Ouroboros.Network.Protocol.VersionNegotiation.Type
+  (
+  -- * Version Negotiation Protocol
+    VersionNegotiationProtocol (..)
+  , Message (..)
+  , ClientHasAgency (..)
+  , ServerHasAgency (..)
+  , NobodyHasAgency (..)
+  , VersionNegotiationError (..)
+
+  -- ** Version Negotiation client
+  , versionNegotiationClientPeer
+  -- ** Version Negotiation server
+  , versionNegotiationServerPeer
+
+  -- ** Tests
+  , pureVersionNegotiation
+  )
+  where
+
+
+import           Data.Typeable (Typeable, cast)
+import           Data.List (intersect)
+import           Data.Map (Map)
+import qualified Data.Map as Map
+
+import qualified Codec.CBOR.Term     as CBOR
+
+import           Network.TypedProtocol.Core
+
+import           Ouroboros.Network.Protocol.VersionNegotiation.Version
+
+-- |
+-- State transitions of the version negotation protocol.
+--
+data VersionNegotiationProtocol vNumber vBlob where
+    StPropose :: VersionNegotiationProtocol vNumber vBlob
+    StConfirm :: VersionNegotiationProtocol vNumber vBlob
+    StDone    :: VersionNegotiationProtocol vNumber vBlob
+
+
+instance Protocol (VersionNegotiationProtocol vNumber vBlob) where
+
+    data Message (VersionNegotiationProtocol vNumber vBlob) from to where
+
+      -- |
+      -- Propose a list of versions together with version data.  It must be
+      -- encoded to a sorted list.
+      --
+      MsgProposeVersions
+        :: Map vNumber vBlob
+        -> Message (VersionNegotiationProtocol vNumber vBlob) StPropose StConfirm
+
+      -- |
+      -- The remote end decide which version to use and sends it back:
+      --
+      MsgAcceptVersion
+        :: vNumber
+        -> vBlob
+        -> Message (VersionNegotiationProtocol vNumber vBlob) StConfirm StDone
+
+      -- |
+      -- or it refuses to run any version,
+      --
+      MsgRefuse
+        :: Message (VersionNegotiationProtocol vNumber vBlob) StConfirm StDone
+
+      -- |
+      -- or a give the client an explicit message that the version was not
+      -- recognised.
+      --
+      MsgUnknownVersion
+        :: Message (VersionNegotiationProtocol vNumber vBlob) StConfirm StDone
```

Comment:
Should include the version number(s) that weren't recognized no?

Reply:
Good idea.  Also, now the server is only checking a single version (largest common version) and it is not trying to decode any other version is something goes wrong.